x-pack/plugin/vector-tile/src/main/java/org/elasticsearch/xpack/vectortile/rest/GridAggregation.java

```diff
@@ -0,0 +1,241 @@
+        // to the density of tile bins but trying not to be bigger.
+        // Level unique tiles H3 resolution unique hexes ratio
+        // 1 4 0 122 30.5
+        // 2 16 0 122 7.625
+        // 3 64 1 842 13.15625
+        // 4 256 1 842 3.2890625
+        // 5 1024 2 5882 5.744140625
+        // 6 4096 2 5882 1.436035156
+        // 7 16384 3 41162 2.512329102
+        // 8 65536 3 41162 0.6280822754
+        // 9 262144 4 288122 1.099098206
+        // 10 1048576 4 288122 0.2747745514
+        // 11 4194304 5 2016842 0.4808526039
+        // 12 16777216 6 14117882 0.8414913416
+        // 13 67108864 6 14117882 0.2103728354
+        // 14 268435456 7 98825162 0.3681524172
+        // 15 1073741824 8 691776122 0.644266719
+        // 16 4294967296 8 691776122 0.1610666797
+        // 17 17179869184 9 4842432842 0.2818666889
+        // 18 68719476736 10 33897029882 0.4932667053
+        // 19 274877906944 11 237279209162 0.8632167343
+        // 20 1099511627776 11 237279209162 0.2158041836
+        // 21 4398046511104 12 1660954464122 0.3776573213
+        // 22 17592186044416 13 11626681248842 0.6609003122
+        // 23 70368744177664 13 11626681248842 0.165225078
+        // 24 281474976710656 14 81386768741882 0.2891438866
+        // 25 1125899906842620 15 569707381193162 0.5060018015
+        // 26 4503599627370500 15 569707381193162 0.1265004504
+        // 27 18014398509482000 15 569707381193162 0.03162511259
+        // 28 72057594037927900 15 569707381193162 0.007906278149
+        // 29 288230376151712000 15 569707381193162 0.001976569537
+        private static final int[] ZOOM2RESOLUTION = new int[] {
+            0,
+            0,
+            0,
+            1,
+            1,
+            2,
+            2,
+            3,
+            3,
+            4,
+            4,
+            5,
+            6,
+            6,
+            7,
+            8,
+            8,
+            9,
+            10,
+            11,
+            11,
+            12,
+            13,
+            13,
+            14,
+            15,
+            15,
+            15,
+            15,
+            15 };
+
+        @Override
+        public GeoGridAggregationBuilder newAgg(String aggName) {
+            return new GeoHexGridAggregationBuilder(aggName);
+        }
+
+        @Override
+        public Rectangle bufferTile(Rectangle tile, int z, int gridPrecision) {
+            if (z == 0 || gridPrecision == 0) {
+                // no need to buffer at level 0 as we are looking to all data.
+                return tile;
+            }
+            final int aggPrecision = gridPrecisionToAggPrecision(z, gridPrecision);
+            if (aggPrecision < 2) {
+                // we need to consider all data
+                return new Rectangle(-180, 180, GeoTileUtils.LATITUDE_MASK, -GeoTileUtils.LATITUDE_MASK);
+            }
+            return new Rectangle(
```

Comment:
Wow, the constructor of Rectangle was confusion. I thought you had made a mistake with the 3rd and 4th arguments swapped around, but when I looked at the code for Rectangle it really does have `minX, maxX, maxY, minY`! I guess this is related to the convention to have the original at the top-left instead of bottom left?

Reply:
Yes, this is legacy and like you I found it confusing.